specs/networking/libp2p-standardization.md

```diff
@@ -141,18 +153,327 @@ The `data` field of a Gossipsub `Message` is an SSZ-encoded object. For the `bea
-its own encryption, ip-discovery and topic advertisement. Therefore, it has no
+is used for discovery. This protocol uses a UDP transport and specifies
+its own encryption, IP-discovery and topic advertisement. Therefore, it has no
 need to establish streams through `multistream-select`, rather, act
 as a standalone implementation that feeds discovered peers/topics (ENR-records) as
-`multiaddrs` into the libp2p service.
+`multiaddrs` into the libp2p service. The libp2p service subsequently forms
+connections and substreams with discovered peers.
+
+## Eth-2 RPC
+
+*There is a specification for the [RPC Interface](./rpc-interface.md) in this
+repository which has implementations for libp2p and non-libp2p client. This
+section focuses solely on the libp2p specification and provides significant
+libp2p-specific modifications - as suggested by @Zah.*
+
+
+### Specification
+
+#### Protocol Segregation
+
+Each RPC message is segregated into it's own libp2p protocol id, which is a string of the form:
+
+```
+<ProtocolPrefix>/<MessageName>/<SchemaVersion>/<Encoding>
+```
+
+With:
+* **ProtocolPrefix** -- the RPC messages are grouped into families identified by a shared LibP2P protocol name prefix. A conforming implementation is expected to support either all messages appearing within a family or none of them. In this case, we use `/eth/serenity/rpc`.
+* **MessageName** -- each RPC request is identified by a name consisting of English letters, digits and underscores (_).
+* **SchemaVersion** -- a semantic version consisting of one or more numbers separated by dots (.). Each schema is versioned to facilitate backward and forward-compatibility when possible.
+* **Encoding** -- while the schema defines the data types in more abstract terms, the encoding describes a specific representation of bytes that will be transmitted over the wire. See the [Encodings](#encodings) section, for further details.
+
+This protocol segregation allows libp2p `multistream-select` to handle the RPC-type, version and encoding negotiation before establishing the underlying substreams.
+
+
+#### Requests and Responses
+
+##### Request
+
+A request is formed by initiating a connection with the protocol id matching
+the desired request type, encoding and version. Once a successful substream is
+negotiated, the request is sent with the matching encoding and prefixed by an
+unsigned varint (as specified by the [protobuf
+docs](https://developers.google.com/protocol-buffers/docs/encoding#varints))
+representing the length of the encoded request (in bytes). More specifically, a typical
+request, looks like:
+
+```
++--------+--------+--------+--------+
+| unsigned_varint | encoded_request |
++--------+--------+--------+--------+
+```
+
+Once a stream is negotiated, the requester SHOULD send the request within **3
+seconds**.
+
+The requester SHOULD then wait for a response on the negotiated stream for at
+most **10 seconds**, before dropping the stream. A requester SHOULD decode the
```

Comment:
Good question! I had intended on the requester to drop their end of the stream, but in this case `Stream.reset()` would be more reasonable. I'm working with my knowledge of rust-libp2p which I don't recall seeing the `reset()` functionality as is likely in go. 
Let me dig a little deeper, perhaps we should update this to `reset()` for clarity.

Reply:
No problem. I'm not 100% sure if we should use reset here though. IMO specifying the stream close/reset will be clearer when we implement the stream handlers. I will also look into it. Thank you. 🙏

Reply:
It's worth noting that libp2p streams are half-open, meaning that in order for the stream to be completely closed, both ends should call `close`, otherwise, the stream remains readable by the sender and writable by the receiver. Not sure about all other implementations but js and go should implement `reset` in the `mplex` muxer.